sram.c

```diff
@@ -0,0 +1,96 @@
+#include "sram.h"
+#include "spi.h"
+#include "nuts_bolts.h"
+
+/* Least and most significant bytes of x, of type uint16_t */
+#define LSB(x) (x & 0x00FF)
+#define MSB(x) ((x & 0xFF00) >> 8)
+
+#define MODE_MASK 0xC0
+#define MODE_IDX  6U
+
+/* SPI insructions */
+enum instruction_e {
+  READ = 0x3,
+  WRITE = 0x2,
+  EDIO = 0x3B,
+  EQIO = 0x38,
+  RSTIO = 0xFF,
+  RDMR = 0x5,
+  WRMR = 0x1
+};
+
+void sram_init()
+{
+  /* Configure SCS pin as output */
+  SCS_SRAM_DDR |= (1 << SCS_SRAM_DDR_PIN);
+
+  /* Set SCS pin to high. The CS for this chip is active low */
+  SCS_SRAM_PORT &= ~(1 << SCS_SRAM_PIN);
+
+  /* SCK resting state is 0. Clock data on rising edge */
+  spi_set_mode(0, 0);
+
+  sram_set_mode(BYTE_MODE);
+}
+
+uint8_t sram_read_mode()
+{
+  spi_set_mode(0, 0);
+
+  uint8_t data_out[2] = {RDMR, 0xFF};
+  uint8_t data_in[2];
+
+  bit_false(SCS_SRAM_PORT, 1 << SCS_SRAM_PIN);
+  spi_transact_array(data_out, data_in, 2);
+  bit_true(SCS_SRAM_PORT, 1 << SCS_SRAM_PIN);
+
+  return (data_in[1] & MODE_MASK) >> MODE_IDX;
+}
+
+void sram_set_mode(enum sram_mode_e mode)
+{
+  spi_set_mode(0, 0);
+
+  uint8_t data_out[2] = {WRMR, mode << MODE_IDX};
+  uint8_t data_in[2];
+
+  bit_false(SCS_SRAM_PORT, 1 << SCS_SRAM_PIN);
+  spi_transact_array(data_out, data_in, 2);
+  bit_true(SCS_SRAM_PORT, 1 << SCS_SRAM_PIN);
+}
+
+uint8_t sram_read_byte(uint16_t addr)
+{
+  spi_set_mode(0, 0);
+
+  uint8_t data_out[4] = {READ,
+                         MSB(addr),
+                         LSB(addr),
+                         0xFF};
```

Comment:
Similarly, `sram_read_byte` and `sram_write_byte` are identical except for two bytes of this array and the return value (one function doesn't return anything). Can they be refactored to reduce code duplication, too? ...would that make semantic sense?

I only half understand this file. I wonder if my suggestions, though well-intentioned, would actually make this code worse.